lib_eio/eio.mli

```diff
@@ -87,18 +103,24 @@ module Switch : sig
       and the switch may have been cancelled by the time it runs. *)
 
   type hook
-  (** A handle for removing a callback. *)
+  (** A handle for removing a clean-up callback. *)
 
   val null_hook : hook
   (** A dummy hook. Removing it does nothing. *)
 
   val on_release_cancellable : t -> (unit -> unit) -> hook
-  (** Like [on_release], but the handler can be removed later. *)
+  (** Like [on_release], but the handler can be removed later.
+
+      For example, opening a file will call [on_release_cancellable] to ensure the file is closed later.
+      However, if the file is manually closed before that, it will use {!remove_hook} to remove the hook,
+      which is no longer needed. *)
```

Comment:
Note: I would find it more natural to have an API like:

```
module Hook : sig
  type t

  val make : (unit -> unit) -> t
  (* [null_hook] is equivalent to [make ignore] *)

  val remove : t -> unit
end

val on_release_cancellable : t -> hook -> unit
(* or just on_release_hook ? *)
```

(There could also be a function to run a hook, to `val run : t -> unit`, but I understand that this is not necessary for internal usage of hooks, only if users want to use them for other things.)

Reply:
I did have it as separate module originally. I got rid of it because it seems distracting in the generated docs (it was the first module you saw when looking at Eio's docs, yet most users won't care about it).

A `make` function won't work, because internally a hook is a `Lwt_dllist.node` in some pre-existing linked list.

Reply:
You could still have `make`, I think? `make` would create a one-element circular dllist, and `on_release_cancellable` would splice this dllist into another dllist. But this looses the property that all hooks are "reachable" from some fiber context. Is this property important? (This separation also makes it possible for users to define their hooks for their own purposes, not to attach them to a context -- if they also have a `run` function. But I don't know if this has any use.)

Reply:
Lwt_dllist doesn't provide a way to do what you want (https://github.com/mirage/lwt-dllist/blob/master/src/lwt_dllist.mli). Adding a node to a list could be useful in some cases, but here it would just add extra complications (what if I try to add a hook that's already in another list?).

odoc just shows things in the same order as the mli file as far as I know. It wouldn't be a huge problem having it as a separate module, but it just doesn't seem important enough for that.